src/Esprima/ParserExtensions.cs

```diff
@@ -15,617 +16,631 @@ static ParserExtensions()
         }
     }
 
-    public static string Slice(this string source, int start, int end)
+    private static string? TryGetInternedString(ReadOnlySpan<char> source)
     {
-        var len = source.Length;
-        var from = start < 0 ? Math.Max(len + start, 0) : Math.Min(start, len);
-        var to = end < 0 ? Math.Max(len + end, 0) : Math.Min(end, len);
-        var span = Math.Max(to - from, 0);
+        var span = source.Length;
 
         // check some common cases
         if (span == 1)
         {
-            return CharToString(source[from]);
+            return CharToString(source[0]);
         }
 
         if (span == 2)
         {
-            if (source[from] == 'i')
+            if (source[0] == 'i')
             {
-                if (source[from + 1] == 'f')
+                if (source[1] == 'f')
                 {
                     return "if";
                 }
 
-                if (source[from + 1] == 'd')
+                if (source[1] == 'd')
                 {
                     return "id";
                 }
 
-                if (source[from + 1] == 'n')
+                if (source[1] == 'n')
                 {
                     return "in";
                 }
             }
 
-            if (source[from] == '"' && source[from + 1] == '"')
+            if (source[0] == '"' && source[1] == '"')
             {
                 return "\"\"";
             }
 
-            if (source[from] == '\'' && source[from + 1] == '\'')
+            if (source[0] == '\'' && source[1] == '\'')
             {
                 return "''";
             }
 
-            if (source[from] == 'f' && source[from + 1] == 'n')
+            if (source[0] == 'f' && source[1] == 'n')
             {
                 return "fn";
             }
 
-            if (source[from] == 'o' && source[from + 1] == 'n')
+            if (source[0] == 'o' && source[1] == 'n')
             {
                 return "on";
             }
 
-            if (source[from] == 'd' && source[from + 1] == 'o')
+            if (source[0] == 'd' && source[1] == 'o')
             {
                 return "do";
             }
         }
         else if (span == 3)
         {
-            if (source[from] == 'a')
+            if (source[0] == 'a')
             {
-                if (source[from + 1] == 'd' && source[from + 2] == 'd')
+                if (source[1] == 'd' && source[2] == 'd')
                 {
                     return "add";
                 }
 
-                if (source[from + 1] == 'r' && source[from + 2] == 'g')
+                if (source[1] == 'r' && source[2] == 'g')
                 {
                     return "arg";
                 }
             }
 
-            if (source[from] == 'f' && source[from + 1] == 'o' && source[from + 2] == 'r')
+            if (source[0] == 'f' && source[1] == 'o' && source[2] == 'r')
```

Comment:
Commenting on this PR though I understand it's unrelated, was just wondering what techniques were used here. 

`Span.IndexOf()` is vectorized now, so checking each char should be slower, and is harder to read.

Reply:
https://github.com/sebastienros/esprima-dotnet/pull/317 may improve the situation.